src/js_native_api_v8.cc

```diff
@@ -355,6 +356,16 @@ class Reference : public RefBase {
     }
   }
 
+ protected:
+  inline void Finalize(bool is_env_teardown = false) override {
+    // During env teardown, `~napi_env()` alone is responsible for finalizing.
+    // Thus, we don't want any stray gc passes to trigger a second call to
+    // `Finalize()`, so let's reset the persistent here.
+    if (is_env_teardown) _persistent.ClearWeak();
```

Comment:
We discussed this solution with @legendecas during today's Node-API meeting and we realized that using `ClearWeak()` to ensure that the engine does not cause `Finalize()` to re-enter by triggering a gc pass during the user's finalizer assumes that `ClearWeak()` causes queued finalizers to be removed from the queue. If this is not the case, it would be better to avoid this re-entrancy by using a flag that would cause `Finalize()` to short-circuit if it was already started as part of env teardown:

```cpp
inline void Finalize(bool is_env_teardown = false) override {
  if (is_env_teardown) _env_teardown_finalize_started = true;
  if (!is_env_teardown && _env_teardown_finalize_started) return;
...
```

where `_env_teardown_finalize_started` is a member variable.